src/Http/Middleware/HandleNotFound.php

```diff
@@ -135,4 +137,30 @@ private function cacheNewRedirect(\Statamic\Sites\Site $site, RedirectContract $
 
         Cache::put('statamic.redirect.redirects', $this->cachedRedirects);
     }
+
+    public function mergeQuery(Request $request, string $destination): string
+    {
+        if (! config('statamic.redirect.preserve_query_strings')) {
+            return $destination;
+        }
+
+        if (! $request->getQueryString()) {
+            return $destination;
+        }
+
+        $destination_parsed = parse_url($destination);
+        $destination_query = [];
+
+        if (isset($destination_parsed['query'])) {
+            parse_str($destination_parsed['query'], $destination_query);
+        }
+
+        $query = array_merge($destination_query, $request->query());
+
+        if (count($query)) {
+            $destination .= '?' . http_build_query($query);
```

Comment:
This can still lead to invalid URLs with duplicate question marks when the destination URL already contains a query string. The query string needs to be removed first before appending the merged query string.

In the test below you are setting a query string on the source URL but not on the destination URL, which is a different scenario.